src/lib.rs

```diff
@@ -476,6 +476,37 @@ pub mod unsync {
             Ok(())
         }
 
+        /// Sets the contents of this cell to `value` and returns a reference to it.
+        ///
+        /// Returns `Ok(&value)` if the cell was empty and `Err(value)` if it was
+        /// full.
+        ///
+        /// # Example
+        /// ```
+        /// use once_cell::unsync::OnceCell;
+        ///
+        /// let cell = OnceCell::new();
+        /// assert!(cell.get().is_none());
+        ///
+        /// assert_eq!(cell.set_and_get(92), Ok(&92));
+        /// assert_eq!(cell.set_and_get(62), Err(62));
+        ///
+        /// assert!(cell.get().is_some());
+        /// ```
+        pub fn set_and_get(&self, value: T) -> Result<&T, T> {
+            let slot = unsafe { &*self.inner.get() };
+            if slot.is_some() {
+                return Err(value);
+            }
+            let slot = unsafe { &mut *self.inner.get() };
+            // This is the only place where we set the slot, no races
+            // due to reentrancy/concurrency are possible, and we've
+            // checked that slot is currently `None`, so this write
+            // maintains the `inner`'s invariant.
+            *slot = Some(value);
+            Ok(slot.as_ref().unwrap())
```

Comment:
It's possible to use `unsafe` to extract the value here instead of using `unwrap`.

Reply:
Like this basically:
```suggestion
            if let Some(value) = slot.as_ref() {
                Ok(value)
            } else {
                unsafe { core::hint::unreachable_unchecked() }
            }
```